### PR TITLE
Implement Elligator2 hash-to-curve for Twisted Edwards curves

### PR DESCRIPTION
**Pull request by drskalman**
## Description

The IETF standard requires/recommond Elligator 2 hash-to-curve for Twisted Edwards Curve. This PR implement the Elligator2 map-to-curve so such hashing is possible in Arkworks.

- [x] Targeted PR against correct branch (master)
- [x] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the GitHub PR explorer

### CODE REVIEW

**Comment by burdges** [comment]
"The IETF standard requires/recommends Elligator 2 hash-to-curve for [for any curve with a ] twisted Edwards form", so ed_on_*, bls12-377, etc.

It won't be easily usable without https://github.com/arkworks-rs/algebra/pull/643 or similar of course, given the current interface is basically exclusive to bls12-381.

**Comment by weikengchen** [comment]
Jeff, I would likely take a look at #643 separately, for the reason that that one is a considerably larger PR...

**Comment by burdges** [comment]
Sure.  Also, there are a few small parts of #643 like https://github.com/arkworks-rs/algebra/pull/643/commits/15fc1a56ccfc8c911372f2411337a0ab04a24138 which permit all of this to be moved out of algebra.  That's not ideal long-term, but it'd simplify using hash-to-curve besides bls12-381 now.

**Comment by drskalman** [comment]
>  given the current interface is basically exclusive to bls12-381.

Hmm I wonder why is that. I have implemented a complete example for a TE curves in the test example. no?

**Comment by drskalman** [comment]
@mmagician you are familiar with the code, it is mostly borrowed from the SWU map. You should be able to review it quite easily.

**Comment by burdges** [comment]
> Hmm I wonder why is that. I have implemented a complete example for a TE curves in the test example. no?

What you have here is fine.  It's not possible to make it the default for a curve.  It's also not possible to generate the base field elements using the correct field hasher either (shake).  #643 already solves those issues.  I've already slurped too much up into #643 though, so it's fine to do this part separately.

**Comment by Pratyush** [comment]
What's the status of this?

**Comment by mmagician** [comment]
I think it's ready, I will review it against the spec.

File: ec/src/hashing/curve_maps/elligator2/mod.rs
Comment (mmagician) [inline ec/src/hashing/curve_maps/elligator2/mod.rs:34]: I would make these sanity checks into `debug_assert` instead.
Actually, I notice that the same can and should be done to `swu` and `wb` maps.

File: ec/src/hashing/curve_maps/elligator2/mod.rs
Comment (mmagician) [inline ec/src/hashing/curve_maps/elligator2/mod.rs:47]:
```suggestion
```

Comment (mmagician) [reply ec/src/hashing/curve_maps/elligator2/mod.rs:47]: I think it's a copy-paste leftover from swu

File: ec/src/hashing/curve_maps/elligator2/mod.rs
Comment (mmagician) [inline ec/src/hashing/curve_maps/elligator2/mod.rs:102]:
```suggestion
        let (x, y): (P::BaseField, P::BaseField) = if gx1.legendre().is_qr() {
            (
                x1,
                gx1.sqrt()
                    .expect("We have checked that gx1 is a quadratic residue. Q.E.D"),
            )
        } else {
            (
                x2,
                gx2.sqrt()
                    .expect("gx2 is a quadratic residue because gx1 is not. Q.E.D"),
            )
        };
```

File: ec/src/hashing/curve_maps/elligator2/mod.rs
Comment (mmagician) [inline ec/src/hashing/curve_maps/elligator2/mod.rs:17]:
```suggestion
/// according to [\[HSSWW23\]]
///
/// - [\[BHKL13\]] <http://dx.doi.org/10.1145/2508859.2516734>
/// - [\[HSSWW23\]] <https://datatracker.ietf.org/doc/html/rfc9380>
```

File: ec/src/hashing/curve_maps/elligator2/mod.rs
Comment (mmagician) [inline ec/src/hashing/curve_maps/elligator2/mod.rs:21]:
```suggestion
    /// When `BaseField` is a prime field, [\[HSSWW23\]] mandates that `Z` is the
```

File: ec/src/hashing/curve_maps/elligator2/mod.rs
Comment (mmagician) [inline ec/src/hashing/curve_maps/elligator2/mod.rs:110]:
```suggestion
        // we just implement the rational map here from [\[HSSWW23\]] Appendix D
```

**Review by mmagician — CHANGES_REQUESTED** [review]

File: ec/src/hashing/curve_maps/elligator2/mod.rs
Comment (mmagician) [inline ec/src/hashing/curve_maps/elligator2/mod.rs:33]:
```suggestion
        // Verifying that Z is a non-square
```

File: ec/src/hashing/curve_maps/elligator2/mod.rs
Comment (mmagician) [inline ec/src/hashing/curve_maps/elligator2/mod.rs:48]: could we be consistent and call it `point` like in the other mappings?

Comment (mmagician) [reply ec/src/hashing/curve_maps/elligator2/mod.rs:48]: Or at least `u` like in RFC 9380

File: ec/src/hashing/curve_maps/elligator2/mod.rs
Comment (mmagician) [reply ec/src/hashing/curve_maps/elligator2/mod.rs:102]: Actually that's only half-right. It seems that parts of steps 6,7 are missing:
```
        // 6. If is_square(gx1), set x = x1, y = sqrt(gx1) with sgn0(y) == 1.
        // 7. Else set x = x2, y = sqrt(gx2) with sgn0(y) == 0.
```
In fact, when I place an assert in the code:
`        assert_eq!(parity(&y), sgn0, "parity of y is not equal to sgn0(y)");`
both tests fail :(

I think that what we need is to return `sgn0`, then call `parity(&y)` and if they mismatch, negate `y`.

Comment (mmagician) [inline ec/src/hashing/curve_maps/elligator2/mod.rs:102]:
```suggestion
        let (x, mut y, sgn0): (P::BaseField, P::BaseField, bool) = if gx1.legendre().is_qr() {
            (
                x1,
                gx1.sqrt()
                    .expect("We have checked that gx1 is a quadratic residue. Q.E.D"),
                true,
            )
        } else {
            (
                x2,
                gx2.sqrt()
                    .expect("gx2 is a quadratic residue because gx1 is not. Q.E.D"),
                false,
            )
        };
        
        if parity(&y) != sgn0 {
            y = -y;
        }
```

**Review by mmagician — CHANGES_REQUESTED** [review]

File: ec/src/hashing/curve_maps/elligator2/mod.rs
Comment (mmagician) [inline ec/src/hashing/curve_maps/elligator2/mod.rs:110]:
```suggestion
        // `(s, t)` is an affine point on the Montgomery curve.
        // Ideally, the TECurve would come with a mapping to its Montgomery curve, 
        // so we could just call that mapping here.
        // This is currently not supported in arkworks, so
        // we just implement the rational map here from [\[HSSWW22\]] Appendix D
```

**Review by mmagician — CHANGES_REQUESTED** [review]

File: ec/src/hashing/curve_maps/elligator2/mod.rs
Comment (Pratyush) [reply ec/src/hashing/curve_maps/elligator2/mod.rs:102]: Do we need the types here?

Comment (drskalman) [reply ec/src/hashing/curve_maps/elligator2/mod.rs:100]: No apparently we don't I'll remove them commiting @mmagician suggestion.

File: ec/src/hashing/curve_maps/elligator2/mod.rs
Comment (drskalman) [reply ec/src/hashing/curve_maps/elligator2/mod.rs:46]: I used point in SWU because pasta implementation used point, but it is not a point per se. We do not impose any topology or geometry on the field, so it is more correct to call it an element (which is going to be mapped to a point on the curve which has geometrical properties). So I found it confusing to call the domain element a point when it is not on a curve. As such, I'm happier to change the name of the parameter in SWU and WB to element instead.

Having said that, if you prefer `point` over `element` I'm not going to dwell on it :-) your call.

Comment (drskalman) [reply ec/src/hashing/curve_maps/elligator2/mod.rs]: Done in https://github.com/arkworks-rs/algebra/pull/659/commits/354db69df8048c6ec6a0027c11a03b7f9d7bf15d

File: ec/src/hashing/curve_maps/elligator2/mod.rs
Comment (drskalman) [reply ec/src/hashing/curve_maps/elligator2/mod.rs]: Done in https://github.com/arkworks-rs/algebra/pull/659/commits/354db69df8048c6ec6a0027c11a03b7f9d7bf15d

**Comment by drskalman** [comment]
> y mention this. Otherwise green light.

Done in https://github.com/arkworks-rs/algebra/pull/659/commits/d8667ca7aa2a9d5800128bea3c87c1e1349f694f

File: ec/src/hashing/curve_maps/elligator2.rs
Comment (Pratyush) [inline ec/src/hashing/curve_maps/elligator2.rs:60]: Can this be precomputed in some way?

File: ec/src/hashing/curve_maps/elligator2.rs
Comment (Pratyush) [inline ec/src/hashing/curve_maps/elligator2.rs:62]: We are computing k.inverse() twice: once right above and once here.

File: ec/src/hashing/curve_maps/elligator2.rs
Comment (drskalman) [reply ec/src/hashing/curve_maps/elligator2.rs:60]: @Pratyush I see you have already brought up issue 1. yourself: https://github.com/arkworks-rs/algebra/pull/627#issuecomment-1520918942

Comment (Pratyush) [reply ec/src/hashing/curve_maps/elligator2.rs:60]: I'd be happy to revert the `new` removal; I assumed y'all had come to some consensus on that front internally at Web3.

Comment (weikengchen) [reply ec/src/hashing/curve_maps/elligator2.rs:60]: So I think the consensus is to resume the `new` so that we have opportunities for precomputation? I agree that this is a good approach because otherwise we need to hardcode the parameters, but these parameters are not super easy to compute: https://github.com/FindoraNetwork/noah/blob/develop/crypto/src/hashing_to_the_curve/ed25519/sswu.rs#L27

Comment (drskalman) [reply ec/src/hashing/curve_maps/elligator2.rs:60]: No we don't have consensus on Web3 side as we have hard-to-convince anti-OOP worriers here. That is why I asked to see what have been on Arkworks reviewers mind to sign off on removing `new` and the ability of having longer-than-one-hash data for MapToCurve.

The counter argument is:

  "This is NOT how substrate+frame works. Frame pretends the lifetime of your code is each tx."

I doubt the validity of the statement (Frame pallets could have longer lived objects) but for the sake of argument suppose it is true.

1. Should Arkworks remove the ability of ever storing any runtime customization/pre-computation by any MapToCurve in future because a specific part of a specific implementation of a specific protocol is not able to take advantage of that optimization. 

2. Even in cases like Elligator where the pre-computation is rather simple and perhaps it makes sense to be outsourced to the curve implementor, it is even more important to enforce verification of the pre-computation in `new`, rather than to leave at the mercy of the same curve implementor to call `check_parameters`. We could feature-decorate the check in `new` so it only happens in debug runs. In this way it won't take an efficiency toll where the code is unable to keep the map alive for more one hash.

Comment (burdges) [reply ec/src/hashing/curve_maps/elligator2.rs:60]: Why is there even a discussion?  `const fn` can do this just fine.  Or do we have some required methods which cannot be made cost?

Comment (Pratyush) [reply ec/src/hashing/curve_maps/elligator2.rs:60]: Trait methods are not currently callable in a `const` context.

Comment (burdges) [reply ec/src/hashing/curve_maps/elligator2.rs:60]: Yeah I just looked that up.  oops.

I'd say make `check_parameters` do this
```
if cfg!(debug_assertions) {
    let correct = whatever_computation
    if Self::thing != correct {
        println!("Incorrect parameter should be {}",correct);
    }
}
```

Comment (drskalman) [reply ec/src/hashing/curve_maps/elligator2.rs:60]: Maybe we should first run some benchmarks before we fighting imaginary perfomance foot guns and presumed bottlenecks. This is one-hash (so no keep-alive benefit) before and after merging @burdge 's changes.
```
     Running benches/my_benchmark.rs (~/code/arkworks-algebra/target/release/deps/my_benchmark-1efd55fa18a21b6d)
elligator - with new    time:   [3.1494 µs 3.1551 µs 3.1618 µs]
elligator - no new      time:   [3.9473 µs 3.9711 µs 4.0002 µs]
```
It seems that we have lost 20+% in performance. I'm not sure why and I don't know exactly what they have changed. 

My point is that we had `new` and the option of having a stateful map, it was worknig. There are plenty of other useful stuff we could spend our time on, rather than implementing future-foot-gun prevention measures resulting in 20% loss of our hash perfomance.

Comment (Pratyush) [reply ec/src/hashing/curve_maps/elligator2.rs:60]: Merged.

**Comment by drskalman** [comment]
This is the benchmarking code:
```
    fn hash_arbitary_string_to_curve_elligator2_keep_alive() {
        let test_elligator2_to_curve_hasher = MapToCurveBasedHasher::<
            Projective<TestElligator2MapToCurveConfig>,
            DefaultFieldHasher<Sha256, 128>,
            Elligator2Map<TestElligator2MapToCurveConfig>,
        >::new(&[1])
        .unwrap();

        for i in 0..100000 {
            let to_be_hashed = format!("{}",i);
            let hash_result = test_elligator2_to_curve_hasher.hash(to_be_hashed.as_bytes()).expect("hash should work for all values");
              assert!(
                 hash_result.is_on_curve(),
                 "hash results into a point off the curve"
             );

        }
    }

    fn hash_arbitary_string_to_curve_elligator2_drop_hasher_after_hash() {

        for i in 0..100000 {
            let test_elligator2_to_curve_hasher = MapToCurveBasedHasher::<
                    Projective<TestElligator2MapToCurveConfig>,
                DefaultFieldHasher<Sha256, 128>,
                Elligator2Map<TestElligator2MapToCurveConfig>,
                >::new(&[1])
                .unwrap();
            let to_be_hashed = format!("{}",i);
            let hash_result = test_elligator2_to_curve_hasher.hash(to_be_hashed.as_bytes()).expect("hash should work for all values");
              assert!(
                 hash_result.is_on_curve(),
                 "hash results into a point off the curve"
             );

        }
    }
```
`cargo bench` automatically compiles in `release` so no debug assert. it could be that computing in stack is faster than computing with `const` somehow.

Sure go ahead and revert the removal of the new and I would revert to 74e07bc.

**Comment by mmagician** [comment]
@drskalman I've re-reverted the necessary [commit](https://github.com/arkworks-rs/algebra/pull/696/), so that we end up without `new` parameters. Once that's merged, let's go ahead and adapt this PR to use `const`, we can then merge this.

**Comment by mmagician** [comment]
@drskalman, the revert PR was just merged. We can now go ahead with the changes here, after adapting to the new API (removed `new`)

**Comment by drskalman** [comment]
@mmagician I removed new and changed the interface. It should be ready now.

**Review by mmagician — APPROVED** [review]
Thanks @drskalman for quickly handling the changes, green light from my side!

File: ec/src/hashing/curve_maps/elligator2.rs
Comment (Pratyush) [inline ec/src/hashing/curve_maps/elligator2.rs:17]:
```suggestion
```

File: ec/src/hashing/curve_maps/elligator2.rs
Comment (Pratyush) [inline ec/src/hashing/curve_maps/elligator2.rs:204]:
```suggestion
```

File: ec/src/hashing/curve_maps/swu.rs
Comment (Pratyush) [inline ec/src/hashing/curve_maps/swu.rs:37]: Should these checks be reverted to `if`s, or do we want to keep them as `debug_assert`s?

File: ec/src/hashing/curve_maps/swu.rs
Comment (Pratyush) [inline ec/src/hashing/curve_maps/swu.rs:89]: What was the motivation of changing this to `debug_assert`?

File: ec/src/hashing/curve_maps/wb.rs
Comment (Pratyush) [inline ec/src/hashing/curve_maps/wb.rs:91]: Ditto

**Review by Pratyush — COMMENTED** [review]
Looks mostly good to me, just a few questions about errors and assertions.

File: ec/src/hashing/curve_maps/swu.rs
Comment (mmagician) [reply ec/src/hashing/curve_maps/swu.rs:37]: This check only concerns the pre-computed parameters, there's nothing runtime dependant.

File: ec/src/hashing/curve_maps/wb.rs
Comment (mmagician) [reply ec/src/hashing/curve_maps/wb.rs:91]: This one also is only dependent on parameters derived at compile time

File: ec/src/hashing/curve_maps/swu.rs
Comment (mmagician) [reply ec/src/hashing/curve_maps/swu.rs:89]: This one is a little more tricky, @drskalman maybe can confirm, but it should never happen at this point that the assert fails right?

File: ec/src/hashing/curve_maps/swu.rs
Comment (mmagician) [inline ec/src/hashing/curve_maps/swu.rs:144]: Here I'm also not sure

File: ec/src/hashing/curve_maps/swu.rs
Comment (mmagician) [reply ec/src/hashing/curve_maps/swu.rs:89]: So, technically under the current code, it could happen that `div = 0`. [Here](https://github.com/arkworks-rs/algebra/pull/659/files#diff-1334780631de1d3814e5a555c6064b485165e7e7b63620c58214ff9209894b7dR73) we apparently get a non-zero `div`, but that's working off the assumption that `P::ZETA` is non-zero.
The current check  on `ZETA` [here](https://github.com/arkworks-rs/algebra/blob/0c945fa0210d70c17f93a42af3e0a25e7e5b757f/ec/src/hashing/curve_maps/swu.rs#L31) should actually be more strict - currently it allows a zero, whereas according to [RFC9380 Sec. 6.6.2](https://www.rfc-editor.org/rfc/rfc9380.pdf): "Z is non-square in F".
With the stricter condition, this `debug_assert` here is fine instead of `assert`.

File: ec/src/hashing/curve_maps/swu.rs
Comment (mmagician) [reply ec/src/hashing/curve_maps/swu.rs:144]: well ok that just follows from the algorithm

File: ec/src/hashing/curve_maps/swu.rs
Comment (mmagician) [inline ec/src/hashing/curve_maps/swu.rs:31]:
```suggestion
            P::ZETA.legendre().is_qnr(),
```

Comment (mmagician) [reply ec/src/hashing/curve_maps/swu.rs:31]: See comment below

File: ec/src/hashing/curve_maps/swu.rs
Comment (Pratyush) [reply ec/src/hashing/curve_maps/swu.rs:89]: I guess my higher-level question was that earlier these checks would happen even in release mode; is it safe for them to happen only in debug mode now?

Comment (drskalman) [reply ec/src/hashing/curve_maps/swu.rs:89]: So it should not fail if the map is correct. So, it is safe if the map is correct. However if an implementor choose a wrong constant, I thought it is a good idea that they get the error that they are mapping off the curve.

However, once they got the paramaters right, then we could forgo the check in the interest of performance assuming that they have at least run it once in the debug mode.
Given that it never fails if the parameters are right, we could remove the check completely if you don't like `debug_assert` rather than slowing done every hashing with `assert`. But I thought we can give a hand to the curve developers in debug time so they can fix their maps.

Your call.

Comment (Pratyush) [reply ec/src/hashing/curve_maps/swu.rs:89]: Makes sense to me, thanks for the explanation!

**Review by Pratyush — APPROVED** [review]

**Pull request merged**
